### ISSUE TITLE
Problems encountered when reading nccl code

### ISSUE THREAD

**Issue by zzjianhui** [NVIDIA/nccl#424]
The following code, what does the variable `width` mean? Get these `widths` in the code, and finally what are they used for. Thank very much!!!

```c
struct ncclTopoLink {
  int type;
  float width; 
  struct ncclTopoNode* remNode; 
};

struct ncclTopoLinkList {
  struct ncclTopoLink* list[NCCL_TOPO_MAX_HOPS];
  int count;
  float width;
  int type;
};
```

**Comment by sjeaugey** [comment]
`width` is the bandwidth in GB/s.

**Comment by zzjianhui** [comment]
@sjeaugey Can you tell me the function of `ncclTopoCompute`?

**Comment by sjeaugey** [comment]
ncclTopoCompute is the graph search. It looks for a maximum number of concurrent paths within the node to maximize the bandwidth of a given pattern (ring or tree). It's a very complex function though as the search may not converge in a reasonable time, so it has to direct the search to the most probably solutions to try to converge quickly.

**Comment by zzjianhui** [comment]
@sjeaugey Thank you for your answer. I have another question. What does the following code mean?
```c
static const double llMaxBws[2][3] = { /* Volta-N1/Intel-N2/Intel-N4) */ {39.0, 39.0, 20.4}, /* Ampere-N1/AMD-N2/AMD-N4) */ {87.7, 22.5 /*avg of ring & tree*/, 19.0} };
```
Can I submit a patch that supports Zhaoxin CPU?

**Comment by sjeaugey** [comment]
The tuning is already way more complex than I'd like it to be. Can you give a bit more context on what you're trying to do, and what kind of problem you encounter? That way we can give a better answer as to how to solve the problem.

**Comment by zzjianhui** [comment]
@sjeaugey Thanks, i have these questions：
You said that the variable `width` in the structure `ncclTopoLink` is the bandwidth,  but the bandwidth calculated according to the pcie formula is different from the code. For example, on pcie3.0x16, the bandwidth is 15.7GB/s, but according to the code calculation is 12GB/s. The following code is the `width` from cpu to cpu. How is this value obtained?

```c
#define QPI_WIDTH 6.0
#define SKL_QPI_WIDTH 9.0
#define P9_WIDTH 32.0
#define ARM_WIDTH 6.0
```

**Comment by sjeaugey** [comment]
> For example, on pcie3.0x16, the bandwidth is 15.7GB/s, but according to the code calculation is 12GB/s. 

These are effective bandwidth for NCCL, not theoretical.

> The following code is the width from cpu to cpu. how to understand this value?

These are what we observe on various CPUs when going across sockets.

> why execute code pciSpeed = INTEL_P2P_OVERHEAD(speed);

Unfortunately CPUs are pretty bad as PCI switches. In particular, Intel CPUs seem to convert all 128B operations to 2x64B requests (which is the cache line size the CPU operates on) even though the request will actually not reach memory but go back to another PCI device. This causes additional inefficiency.

**Comment by zzjianhui** [comment]
@sjeaugey Through Wikipedia, the QPI speed has reached 25.6 GB/s. Assuming that NVLINK is not used on my machine, is this variable important?

**Comment by sjeaugey** [comment]
Yes this variable can be important depending on how much bandwidth we can get through the CPU and between sockets. We don't have special code for the Zhaoxin CPU so the bandwidth should be the default, `LOC_WIDTH` (5000 ~= infinity).

https://github.com/NVIDIA/nccl/blob/master/src/graph/topo.cc#L62-L76

You can see what topology we detect by setting `NCCL_DEBUG=INFO NCCL_DEBUG_SUBSYS=GRAPH`. Also, feel free to dump the topology of your system with `NCCL_TOPO_DUMP_FILE=system.txt` and post the file here, so that we can take a look.

**Comment by zzjianhui** [comment]
Hi @sjeaugey , from Wikipedia, the speed of qpi can reach 25.6GB/s, but here it is only set to 6 or 9 (x86). Whether the two values are too different?
```c
#define QPI_WIDTH 6.0
#define SKL_QPI_WIDTH 9.0
#define P9_WIDTH 32.0
#define ARM_WIDTH 6.0
```

**Comment by sjeaugey** [comment]
The values here are just what we observed. Why we get such low value compared to what should be is a good question which I don't have the answer for.

**Comment by zzjianhui** [comment]
Hi @sjeaugey , have you solved this problem(qpi width value)? If it is resolved, please let me know, thank you very much!

**Comment by sjeaugey** [comment]
The fact some CPUs are not able to achieve peak bandwidth is unfortunate, but not something we can do anything about. We can only note the performance restriction. Our code would be much simpler if all CPUs were achieving peak bandwidth in all cases.

**Comment by andrei-pokrovsky** [comment]
Wouldn't this create a problem with EPYC? Why is the default infinity and not some small number?

static ncclResult_t ncclTopoGetInterCpuWidth(struct ncclTopoNode* cpu, float* width) {
  *width = LOC_WIDTH;
  if (cpu->cpu.arch == NCCL_TOPO_CPU_ARCH_POWER) {
    *width = P9_WIDTH;
    return ncclSuccess;
  }
  if (cpu->cpu.arch == NCCL_TOPO_CPU_ARCH_ARM) {
    *width = ARM_WIDTH;
    return ncclSuccess;
  }
  if (cpu->cpu.arch == NCCL_TOPO_CPU_ARCH_X86 && cpu->cpu.vendor == NCCL_TOPO_CPU_VENDOR_INTEL) {
    *width = cpu->cpu.model == NCCL_TOPO_CPU_TYPE_SKL ? SKL_QPI_WIDTH : QPI_WIDTH;
  }
  if (cpu->cpu.arch == NCCL_TOPO_CPU_ARCH_X86 && cpu->cpu.vendor == NCCL_TOPO_CPU_VENDOR_ZHAOXIN) {
    *width = cpu->cpu.model ==  NCCL_TOPO_CPU_TYPE_YONGFENG ? YONGFENG_ZPI_WIDTH : ZPI_WIDTH;
  }
  return ncclSuccess;
}

**Comment by sjeaugey** [comment]
> Wouldn't this create a problem with EPYC?
True, we haven't added a special rule for EPYC yet, and it could cause issues in some case, so we'll probably want to add one at some point.

**Comment by zfy3000163** [comment]
#define INTEL_P2P(speed) (speed*9/12)
#define INTEL_P2P_OVERHEAD(speed) (speed*12/9)  
@sjeaugey   I have a question, will this setting affect performance? #

**Comment by zfy3000163** [comment]
We found that the bandwidth of QPI is actually sufficient. Why do we need to make these settings for cross-slot transmission?

**Comment by stephenmsachs** [comment]
### Issue Cleanup: Helping Us Focus on Current Challenges

  We're reviewing older issues to ensure we prioritize the most relevant and active ones. Since this issue hasn't seen updates in over 6 months, we'll be closing it for now.

  ,*This change helps us focus our efforts on addressing any current issues our users are facing.* If this issue still affects you, please don't hesitate to reopen it with a quick update (e.g., "Still relevant on [version=X]").
  Thanks for your understanding and for contributing to NCCL.
